### ISSUE TITLE
[Spec] <declaration> production seems to be missing some types of modifiers

### ISSUE THREAD

**Issue by kaby76** [dart-lang/language#2282]
This code can be compiled by the Dart compiler, but it can't be parse with the given rules in the Spec:
```
abstract class StreamController {
	final int i1 = 0;
	abstract int i2; // Can't parse
	abstract final int i3;  // Can't parse
}
```
It fails to parse the declaration for i2 and i3 because of the modifier "abstract".

From the Spec, the rule for [declaration](https://github.com/dart-lang/language/blob/812f939ec0130da119ad2cbe5f9a4121e31b0831/specification/dartLangSpec.tex#L2423) allows for a number of cases, but does not allow for a declaration that begins with `abstract`.

Note, for a class member declaration, the parser itself seems to recognize [three groups of modifiers](https://github.com/dart-lang/sdk/blob/81584f5d031fc12049f954b8510de7c36f1e56ea/pkg/_fe_analyzer_shared/lib/src/parser/parser_impl.dart#L3918-L3977) followed by a [list of modifiers](), roughly as ('external'|'augment'|'abstract')? ('static'|'covariant')? ('final'|'var'|'const'|('late' 'final'?)? [classMemberModifiers](https://github.com/dart-lang/sdk/blob/81584f5d031fc12049f954b8510de7c36f1e56ea/pkg/_fe_analyzer_shared/lib/src/parser/parser_impl.dart#L3965)?; [classMemberModifiers](https://github.com/dart-lang/sdk/blob/81584f5d031fc12049f954b8510de7c36f1e56ea/pkg/_fe_analyzer_shared/lib/src/parser/modifier_context.dart#L96) : [modifiers](https://github.com/dart-lang/sdk/blob/81584f5d031fc12049f954b8510de7c36f1e56ea/pkg/_fe_analyzer_shared/lib/src/parser/modifier_context.dart#L97); [modifiers](https://github.com/dart-lang/sdk/blob/81584f5d031fc12049f954b8510de7c36f1e56ea/pkg/_fe_analyzer_shared/lib/src/parser/modifier_context.dart#L245) : (abstract|augment|const|covariant|external|final|late|required|static|var)*. Anyway, the CFG in the Spec needs alts to allow `abstract` members of a class

**Comment by eernstg** [comment]
External variables (top-level, static, instance) and abstract instance variables were added with null safety (that is, Dart 2.12). The specification updates about null safety are currently being reviewed, so that's the reason why they are not included in the current version of the language specification.

**Comment by eernstg** [comment]
Note that the Dart language specification is being updated based on so-called feature specifications. A feature specification is a document which is similar in style to the language specification (some of them are more formal/precise, others a bit less), and it is used to hold the information about the intended syntax and semantics of an upcoming language feature. Checking out [this directory](https://github.com/dart-lang/language/tree/master/accepted) to find some feature specifications. In particular, the null safety feature specification is [here](https://github.com/dart-lang/language/blob/master/accepted/2.12/nnbd/feature-specification.md), and the feature specification about abstract and external fields is [here](https://github.com/dart-lang/language/blob/master/accepted/2.12/abstract-external-fields/feature-specification.md).

**Comment by kaby76** [comment]
Sorry, I didn't understand your development process. Thanks for explaining.

From experimentation, I gathered [the "2.15-dev" spec](https://github.com/dart-lang/language/blob/812f939ec0130da119ad2cbe5f9a4121e31b0831/specification/dartLangSpec.tex) is a backward-looking document. Yes, it's work-in-progress, and it doesn't yet reflect [the state of the compiler at tag "2.15.1"](https://github.com/dart-lang/sdk/tree/2.15.1/pkg/_fe_analyzer_shared/lib/src/parser).

I extended (fudged) enough of the CFG scraped from [the spec](https://github.com/dart-lang/language/blob/812f939ec0130da119ad2cbe5f9a4121e31b0831/specification/dartLangSpec.tex) to construct an Antlr4 parser that recognizes all but a couple of files from [a significant subset of the sdk](https://github.com/dart-lang/sdk/tree/2.15.1/sdk/lib). More importantly, I can now parse the compiler source itself. I will be using that to bootstrap the CFG directly from the source code.

**Comment by eernstg** [comment]
Right, the version number 2.15 should actually be changed to 2.13 at this point, because it's taken a long time to work on the null safety updates, and we're actually more likely to release a 2.13 spec at this point.

> I extended (fudged) enough of the CFG scraped from [the spec](https://github.com/dart-lang/language/blob/812f939ec0130da119ad2cbe5f9a4121e31b0831/specification/dartLangSpec.tex) to construct an Antlr4 parser

Sounds good! I was wondering why you weren't just using `Dart.g`. If you do find faults in there we should just fix them, and that grammar isn't tied to the same level of scrutiny as the language specification text (so we can move faster, and we can make the choice to model a language version which is basically as recent as we'd want, and it's already close to that).
